actionpack/lib/action_controller/metal/request_forgery_protection.rb

```diff
@@ -446,8 +450,8 @@ def valid_per_form_csrf_token?(token, session) # :doc:
       end
 
       def real_csrf_token(session) # :doc:
-        session[:_csrf_token] ||= generate_csrf_token
-        decode_csrf_token(session[:_csrf_token])
+        csrf_token = csrf_token_registrar.call(request, method(:generate_csrf_token))
```

Comment:
Why we are passing the method and not the value? To avoid generating the token where it is not necessary to set?

In that case maybe using a block here would be better?

```
csrf_token_registry.fetch { generate_csrf_token }
```

Reply:
Yes.  I didn't want to regenerate a new token on every request if it wasn't necessary.  Now that I see it, I totally agree that the block would have been a better option.

I've reworked it to use an object instead of a lambda as requested, and I think it resolves that issue.